File: generator/src/links.rs
use crate::domain::LinkKind;
use url::Url;

pub fn classify_url(raw: &str) -> Option<LinkKind> {
    if let Some(video_id) = parse_youtube_url(raw) {
        return Some(LinkKind::YouTube { video_id });
    }
    let url = Url::parse(raw).ok()?;
    match url.scheme() {
        "http" | "https" => Some(LinkKind::ExternalUrl {
            url: url.to_string(),
        }),
        _ => None,
    }
}

pub fn parse_youtube_url(raw: &str) -> Option<String> {
    let url = Url::parse(raw).ok()?;
    if url.scheme() != "http" && url.scheme() != "https" {
        return None;
    }
    let host = url.host_str()?.to_ascii_lowercase();
    let segments: Vec<_> = url.path_segments().map(|s| s.collect()).unwrap_or_default();

    let candidate = match host.as_str() {
        "youtube.com" | "www.youtube.com" | "m.youtube.com" => {
            if url.path() == "/watch" {
                url.query_pairs()
                    .find(|(k, _)| k == "v")
                    .map(|(_, v)| v.into_owned())
            } else if segments.len() >= 2 && (segments[0] == "shorts" || segments[0] == "embed") {
                Some(segments[1].to_string())
            } else {
                None
            }
        }
        "youtu.be" => segments.first().map(|id| id.to_string()),
        _ => None,
    }?;

    is_valid_youtube_id(&candidate).then_some(candidate)
}

fn is_valid_youtube_id(id: &str) -> bool {
    id.len() == 11
        && id
            .bytes()
            .all(|b| b.is_ascii_alphanumeric() || b == b'-' || b == b'_')
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_supported_youtube_urls() {
        let id = "dQw4w9WgXcQ";
        for url in [
            format!("https://youtube.com/watch?v={id}"),
            format!("https://www.youtube.com/watch?v={id}&t=10"),
            format!("https://m.youtube.com/watch?v={id}"),
            format!("https://youtu.be/{id}?t=10"),
            format!("https://youtube.com/shorts/{id}"),
            format!("https://www.youtube.com/embed/{id}"),
        ] {
            assert_eq!(parse_youtube_url(&url), Some(id.to_string()));
        }
    }

    #[test]
    fn rejects_deceptive_or_malformed_youtube_urls() {
        for url in [
            "https://youtube.com.evil.example/watch?v=dQw4w9WgXcQ",
            "javascript:alert(1)",
            "https://youtube.com/watch?v=short",
            "https://example.com/embed/dQw4w9WgXcQ",
        ] {
            assert_eq!(parse_youtube_url(url), None);
        }
    }
}

File: generator/src/links.ts
import type { BoardLink, PdfDocument } from "./domain.js";
import { pdfToBoardCoordinates } from "./geometry.js";

const YOUTUBE_ID = /^[A-Za-z0-9_-]{11}$/;

type ClassifiedLink =
  | Readonly<{ kind: "youtube"; videoId: string }>
  | Readonly<{ kind: "externalUrl"; url: string }>;

export function parseYoutubeUrl(raw: string): string | null {
  const url = parseHttpUrl(raw);
  if (!url) return null;

  const host = url.hostname.toLowerCase();
  const segments = url.pathname.split("/").filter(Boolean);
  let candidate: string | null = null;
  if (["youtube.com", "www.youtube.com", "m.youtube.com"].includes(host)) {
    candidate = youtubeComId(url, segments);
  } else if (host === "youtu.be") {
    candidate = segments[0] ?? null;
  }
  return candidate && YOUTUBE_ID.test(candidate) ? candidate : null;
}

export function boardLinks(document: PdfDocument): BoardLink[] {
  return document.rawLinks.map((annotation) => {
    const rect = pdfToBoardCoordinates(annotation.rect, document.geometry);
    if (!rect) throw new Error("invalid annotation rectangle");
    const link = classifyUrl(annotation.uri);
    if (!link) throw new Error(`unsafe or unsupported URL: ${annotation.uri}`);
    return { ...rounded(rect), ...link };
  });
}

function classifyUrl(raw: string): ClassifiedLink | null {
  const videoId = parseYoutubeUrl(raw);
  if (videoId) return { kind: "youtube", videoId };
  const url = parseHttpUrl(raw);
  return url ? { kind: "externalUrl", url: url.href } : null;
}

function youtubeComId(url: URL, segments: readonly string[]): string | null {
  if (url.pathname === "/watch") return url.searchParams.get("v");
  return segments.length >= 2 && (segments[0] === "shorts" || segments[0] === "embed")
    ? segments[1] ?? null
    : null;
}

function parseHttpUrl(raw: string): URL | null {
  try {
    const url = new URL(raw);
    return url.protocol === "http:" || url.protocol === "https:" ? url : null;
  } catch {
    return null;
  }
}

function rounded(rect: Readonly<{ x: number; y: number; width: number; height: number }>) {
  const round = (value: number) => Math.round(value * 1000) / 1000;
  return { x: round(rect.x), y: round(rect.y), width: round(rect.width), height: round(rect.height) };
}
